File: Firm Dynamics/Hopenhayn1992.py
# imports
import time
import numpy as np
import scipy as sp
import quantecon as qe
import matplotlib.pyplot as plt
import matplotlib.ticker as mtick
import seaborn as sns
sns.set_theme(style='whitegrid')

class Hopenhayn1992:
    """
    Class object of the Hopenhayn (1992) model.
    """

    #################
    # 1. Initiation #
    #################

    def __init__(self,
                 β = 0.96,      # discount factor
                 B = 100,       # demand primitive
                 α = 2/3,       # labor factor
                 c_f = 15,      # fixed cost
                 c_e = 50,      # entry cost
                 μ_s = 1.2,     # loc of initial productivity draw
                 ρ = 0.9,       # productivity autocorrelation
                 σ_ϵ = 0.2,     # shock coefficient
                 Nz = 20,       # number of discrete states
                 display_plots=True,
                 ):
        self.β, self.B, self.α, self.c_f, self.c_e, self.μ_s, self.ρ, self.σ_ϵ, self.Nz = β, B, α, c_f, c_e, μ_s, ρ, σ_ϵ, Nz
        self.display_plots = display_plots
        self.price_ss = None
        self.setup_parameters()
        self.setup_grid()


    def setup_parameters(self):
        # incumbent firm solution
        self.tol = 1e-8
        self.maxit = 2_000


    def setup_grid(self):
        # discretely approximate the continuous AR(1) process 
        self.mc = qe.markov.approximation.rouwenhorst(n=self.Nz, rho=self.ρ, sigma=self.σ_ε, mu=self.μ_s*(1-self.ρ))

        # transition matrix and states
        self.F = self.mc.P
        self.prod_grid = np.exp(self.mc.state_values)

        # initial productivity distribution for entrant firm
        """ self.initial = qe.markov.approximation.rouwenhorst(n=self.Nz, rho=0, sigma=(self.σ_ε**2)/(1-self.ρ**2), mu=self.μ_s)
        self.G = np.array(self.initial.stationary_distributions)
        self.G = np.squeeze(np.asarray(self.G))
        self.grid_x = np.exp(self.initial.state_values) """

        self.G = np.array(self.mc.stationary_distributions)
        self.G = np.squeeze(np.asarray(self.G))


    def interpol(self,x,y,x1):
        y1 = np.interp(x1,x,y)
        return y1


    def static_profit_max(self, price):
        optimal_n = (self.α * price * self.prod_grid) ** (1 / (1 - self.α))
        firm_output = self.prod_grid * (optimal_n ** self.α)
        firm_profit = price*firm_output - optimal_n - price*self.c_f
        
        return firm_profit, firm_output, optimal_n


    def incumbent_firm(self, price):
        """
        Value function iteration for the incumbent firm problem.
        """ 
        VF_old = np.zeros(self.Nz)
        VF = np.zeros(self.Nz)
        
        firm_profit, firm_output, pol_n = self.static_profit_max(price)
    
        for it in range(self.maxit):
            
            VF = firm_profit + self.β * np.maximum(np.dot(self.F, VF_old),0)
            
            dist = np.abs(VF_old - VF).max()
        
            if dist < self.tol :
               break
           
            VF_old = np.copy(VF)

        continue_indicator = np.ones(self.Nz)*(np.dot(self.F, VF)>=0)
        
        idx = np.searchsorted(continue_indicator, 1) #index of self.pol_continue closest to one on the left
        exit_cutoff = self.prod_grid[idx]
        
        return VF, firm_profit, firm_output, pol_n, continue_indicator, exit_cutoff

    #################
    # 2. Price Loop #
    #################

    def find_equilibrium_price(self):
        """
        Finds the equilibrium price that clears markets. 
        
        The function follows steps 1-3 in algorithm using the bisection method. It guesses a price, solves incumbent firm vf 
        then checks whether free entry condition is satisfied. If not, updates the price and tries again. The free entry condition 
        is where the firm value of the entrant (VF_entrant) equals the cost of entry (ce) and hence the difference between the two is zero. 
        """
        
        pmin, pmax = 0, 100
        for it_p in range(self.maxit):
            price = (pmin+pmax)/2
            
            VF = self.incumbent_firm(price)[0]
        
            VF_entrant = np.dot(VF, self.G)
            
            diff = np.abs(VF_entrant-(price*self.c_e))
            
            if diff < self.tol:
                break
            
            if VF_entrant < price*self.c_e :
                pmin=price 
            else:
                pmax=price
        
        return price
    
    ########################
    # 2. Find Distribution #
    ########################

    def solve_invariant_distribution(self, m, continue_indicator):
        F_tilde = (self.F * continue_indicator.reshape(self.Nz, 1)).T
        I = np.eye(self.Nz)
         
        return m * ( np.dot( np.linalg.inv(I - F_tilde), self.G ) )

    ########################
    # 4. Further Inference #
    ########################
    def track_cohort(self, n):
        exit_rate = np.zeros(shape=n)
        relative_size = np.zeros(shape=n+1)
        for age in range(n+1):
            if age == 0:
                current_dist = self.G
                current_size = self.m_star
                current_gamma = current_dist * current_size
            else:
                current_gamma = self.F.T @ (np.multiply(current_gamma, self.continue_indicator))
                current_size = np.sum(current_gamma)
                current_dist = current_gamma / current_size
            
            current_average_size = np.dot(self.pol_n, current_dist)
            relative_size[age] = current_average_size / self.average_firm_size
            current_continue_rate = np.dot(current_dist,self.continue_indicator)
            if age > 0:
                exit_rate[age-1] = 1 - current_continue_rate

            print(f"Age: {age}")
            if age > 0:
                print(f"{'Exit rate ':.<40}{exit_rate[age-1]:0.4f}")
            print(f"{'Relative firm size ':.<40}{relative_size[age]:0.4f}")
        if self.display_plots == True:
            fig, axes = plt.subplots(1,2,figsize=(10,5))
            axes[0].plot(range(1,n+1), exit_rate)
            axes[0].yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1,decimals=None))
            axes[1].plot(range(n+1), relative_size)
            plt.show()
    

    def solve_model(self):
        """
        Finds the stationary equilibrium
        """  
        
        t0 = time.time()    #start the clock
        
        # Find the optimal price using bisection (algo steps 1-3)
        self.price_ss = self.find_equilibrium_price()
        
        # Use the equilibrium price to recover incumbent firm solution
        self.VF, self.firm_profit, self.firm_output, self.pol_n, self.continue_indicator, self.exit_cutoff = self.incumbent_firm(self.price_ss)
        
        # c. Invariant (productivity) distribution with endogenous exit. Here assume m=1 which 
        #will come in handy in the next step.
        self.gamma_at_1 = self.solve_invariant_distribution(1, self.continue_indicator)
        
        # d. Rather than iterating on market clearing condition to find the equilibrium mass of entrants (m_star)
        # we can compute it analytically (Edmond's notes ch. 3 pg. 25)
        self.Y = self.B / self.price_ss
        self.m_star = self.Y / (np.dot(self.gamma_at_1,self.firm_output - self.c_f) - self.c_e)
        
        # e. Rescale to get invariant (productivity) distribution (mass of plants)
        self.gamma = self.m_star * self.gamma_at_1
        self.total_mass = np.sum(self.gamma)
        
        # Invariant (productivity) distribution by percent
        self.pdf_stationary = self.gamma / self.total_mass
        self.cdf_stationary = np.cumsum(self.pdf_stationary)
        
        # f. calculate employment distributions
        self.distrib_emp = (self.pol_n * self.gamma)
        
        # invariant employment distribution by percent
        self.pdf_emp = self.distrib_emp / np.sum(self.distrib_emp)
        self.cdf_emp = np.cumsum(self.pdf_emp)
        
        # g. calculate statistics
        self.total_employment = np.dot(self.pol_n, self.gamma)
        self.average_firm_size = self.total_employment / self.total_mass
        self.average_entrant_size = np.dot(self.pol_n, self.G)
        self.relative_size_of_new_entrants = self.average_entrant_size / self.average_firm_size
        self.exit_rate = self.m_star / self.total_mass
        #self.exit_rate = 1-(np.sum(self.pi.T*self.distrib_stationary_0*self.pol_continue)/np.sum(self.distrib_stationary_0)) #alternative calculation

        self.report_dict = {
            "Steady State Equilibrium Price": self.price_ss,
            "Entry Mass": self.m_star,
            "Mass of all firms": self.total_mass,
            "Firm entry/exit rate": self.exit_rate,
            "Average Firm Size": self.average_firm_size,
            "Relative Size of New Entrants": self.relative_size_of_new_entrants,
            "Aggregate Output": self.Y,
            "Aggregate Employment": self.total_employment
            }
        
        print("\n-----------------------------------------------------------")
        for string, var in self.report_dict.items():
            print(f"{string:.<40} {var:0.4f}")
        print("-----------------------------------------------------------")
        self.track_cohort(5)

        # h. plot
        
        if self.display_plots:
            plt.plot(self.prod_grid, self.VF)
            plt.axvline(self.exit_cutoff, color='tab:red', linestyle='--', alpha=0.7)
            plt.axhline(0, color='tab:green', linestyle='--', alpha=0.7)
            plt.title('Incumbant Firm Value Function')
            plt.legend(['Value Function', 'Exit Threshold='+str(self.exit_cutoff.round(2))])
            plt.xlabel('Productivity level')
            #plt.savefig('value_func_hopehayn.pdf')
            plt.show()
         
            plt.plot(self.prod_grid,self.pdf_stationary)
            plt.plot(self.prod_grid, self.pdf_emp)
            plt.title('Stationary PDF' )
            plt.xlabel('Productivity level')
            plt.ylabel('Density')
            plt.legend(['Share of Firms','Share of Employment'])
            #plt.savefig('pdf_hopehayn.pdf')
            plt.show()
            
            plt.plot(self.prod_grid,self.cdf_stationary)
            plt.plot(self.prod_grid, self.cdf_emp)
            plt.title('Stationary CDF' )
            plt.xlabel('Productivity level')
            plt.ylabel('Cumulative Sum')
            plt.legend(['Share of Firms','Share of Employment'])
            #plt.savefig('cdf_hopehayn.pdf')
            plt.show()       
        
        t1 = time.time()
        print(f'\nTotal Run Time: {t1-t0:.2f} seconds')

# model = Hopenhayn1992(display_plots=False)
# model.solve_model()

File: Firm Dynamics/Hopenhayn1992.jl
using LinearAlgebra
using Distributions
using Plots
using QuantEcon

mutable struct Hopenhayn1992
    β
    B
    α
    c_f
    c_e
    μ_s
    ρ
    σ_ϵ
    Nz
    display_plots
    price_ss
    tol
    maxit
    F
    prod_grid
    G
    VF
    firm_profit
    firm_output
    pol_n
    continue_indicator
    exit_cutoff
    gamma_at_1
    Y
    m_star
    gamma
    total_mass
    pdf_stationary
    cdf_stationary
    distrib_emp
    pdf_emp
    cdf_emp
    total_employment
    average_firm_size
    average_entrant_size
    relative_size_of_new_entrants
    exit_rate
    report_dict
end

function Hopenhayn1992(;
    β=0.96,
    B=100,
    α=2//3,
    c_f=15,
    c_e=50,
    μ_s=1.2,
    ρ=0.9,
    σ_ϵ=0.2,
    Nz=20,
    display_plots=true,
    tol=1e-8,
    maxit=2000,
    )

    model = Hopenhayn1992(
    β,
    B,
    α,
    c_f,
    c_e,
    μ_s,
    ρ,
    σ_ϵ,
    Nz,
    display_plots,
    nothing,
    tol,
    maxit,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    nothing,
    )
    setup_grid!(model)
    return model
end

function setup_grid!(model::Hopenhayn1992)
    # Discretely approximate the continuous AR(1) process
    mc = rouwenhorst(model.Nz, model.ρ, model.σ_ϵ, model.μ_s * (1 - model.ρ))
    model.F = mc.p
    model.prod_grid = exp.(mc.state_values)
    model.G = stationary_distributions(mc)[1]
end

function static_profit_max(model::Hopenhayn1992, price)
    optimal_n = (model.α * price .* model.prod_grid) .^ (1 / (1 - model.α))
    firm_output = model.prod_grid .* (optimal_n .^ model.α)
    firm_profit = price .* firm_output .- optimal_n .- price .* model.c_f
    return firm_profit, firm_output, optimal_n
end

function incumbent_firm(model::Hopenhayn1992, price)
    VF_old = zeros(model.Nz)
    VF = zeros(model.Nz)
    firm_profit, firm_output, pol_n = static_profit_max(model, price)
    for it in 1:model.maxit
        VF = firm_profit .+ model.β .* max.(model.F * VF_old, 0)
        dist = maximum(abs.(VF_old - VF))
        if dist < model.tol
            break
        end
        VF_old .= VF
    end
    continue_indicator = (model.F * VF .>= 0)
    idx = searchsortedfirst(continue_indicator, true)
    exit_cutoff = model.prod_grid[idx]
    return VF, firm_profit, firm_output, pol_n, continue_indicator, exit_cutoff
end

function find_equilibrium_price(model::Hopenhayn1992)
    pmin, pmax = 0.0, 100.0
    price = 0.0
    for it_p in 1:model.maxit
        price = (pmin + pmax) / 2
        VF = incumbent_firm(model, price)[1]
        VF_entrant = dot(VF, model.G)
        diff = abs(VF_entrant - (price * model.c_e))
        if diff < model.tol
            break
        end
        if VF_entrant < price * model.c_e
            pmin = price
        else
            pmax = price
        end
    end
    return price
end

function solve_invariant_distribution(model::Hopenhayn1992, m, continue_indicator)
    F_tilde = (model.F .* continue_indicator')'
    I_matrix  = I(model.Nz)
    return m .* (inv(I_matrix  - F_tilde) * model.G)
end

function track_cohort(model::Hopenhayn1992, n)
    exit_rate = zeros(n)
    relative_size = zeros(n + 1)
    current_dist = model.G
    current_size = model.m_star
    current_gamma = current_dist .* current_size

    for age in 0:n
        if age > 0
            current_gamma = model.F' * (current_gamma .* model.continue_indicator)
            current_size = sum(current_gamma)
            current_dist = current_gamma ./ current_size
        end

        current_average_size = dot(model.pol_n, current_dist)
        relative_size[age + 1] = current_average_size / model.average_firm_size
        current_continue_rate = dot(current_dist, model.continue_indicator)

        if age > 0
            exit_rate[age] = 1 - current_continue_rate
        end

        println("Age: $age")
        if age > 0
            println("Exit rate: $(exit_rate[age])")
        end
        println("Relative firm size: $(relative_size[age + 1])")
    end

    if model.display_plots
        fig = plot(layout = (1, 2), size = (1000, 500))
    
        # First plot
        plot!(fig[1], 1:n, exit_rate, label = "Exit Rate")
        yformatter = y -> string(round(y * 100, digits = 2)) * "%"
        ylims!(fig[1], 0, 1)
        # yticks!(fig[1], 0:0.1:1)
        # yticklabels!(fig[1], yformatter.(0:0.1:1))
    
        # Second plot
        plot!(fig[2], 0:n, relative_size, label = "Relative Size")
    
        display(fig)
    end
end

function solve_model!(model::Hopenhayn1992)
    t0 = time()
    model.price_ss = find_equilibrium_price(model)
    model.VF, model.firm_profit, model.firm_output, model.pol_n, model.continue_indicator, model.exit_cutoff = incumbent_firm(model, model.price_ss)
    model.gamma_at_1 = solve_invariant_distribution(model, 1, model.continue_indicator)
    model.Y = model.B / model.price_ss
    model.m_star = model.Y / (dot(model.gamma_at_1, model.firm_output .- model.c_f) - model.c_e)
    model.gamma = model.m_star .* model.gamma_at_1
    model.total_mass = sum(model.gamma)
    model.pdf_stationary = model.gamma ./ model.total_mass
    model.cdf_stationary = cumsum(model.pdf_stationary)
    model.distrib_emp = model.pol_n .* model.gamma
    model.pdf_emp = model.distrib_emp ./ sum(model.distrib_emp)
    model.cdf_emp = cumsum(model.pdf_emp)
    model.total_employment = dot(model.pol_n, model.gamma)
    model.average_firm_size = model.total_employment / model.total_mass
    model.average_entrant_size = dot(model.pol_n, model.G)
    model.relative_size_of_new_entrants = model.average_entrant_size / model.average_firm_size
    model.exit_rate = model.m_star / model.total_mass
    model.report_dict = Dict(
        "Steady State Equilibrium Price" => model.price_ss,
        "Entry Mass" => model.m_star,
        "Mass of all firms" => model.total_mass,
        "Firm entry/exit rate" => model.exit_rate,
        "Average Firm Size" => model.average_firm_size,
        "Relative Size of New Entrants" => model.relative_size_of_new_entrants,
        "Aggregate Output" => model.Y,
        "Aggregate Employment" => model.total_employment
    )
    println("\n-----------------------------------------------------------")
    for (string, var) in model.report_dict
        println("$string: $var")
    end
    println("-----------------------------------------------------------")
    track_cohort(model, 5)
    if model.display_plots
        p = plot(model.prod_grid, model.VF, label="Value Function")
        vline!(p, [model.exit_cutoff], label="Exit Threshold")
        hline!(p, [0], label="Zero Line")
        title!(p, "Incumbent Firm Value Function")
        xlabel!(p, "Productivity level")
        display(p)
    end
    t1 = time()
    println("\nTotal Run Time: $(t1 - t0) seconds")
end

# model = Hopenhayn1992()
# @time solve_model!(model)